### PR TITLE
modulo '1' is always zero

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/5822

### CODE REVIEW

File: src/Analyser/MutatingScope.php [hunk 1]
```diff
@@ -1051,12 +1051,24 @@ private function resolveType(Expr $node): Type
...
 				$right = $node->right;
 			}
+			$rightType = $this->getType($right);
+
+			$stringType = $rightType->toString();
+			if (
+				$node instanceof Node\Expr\BinaryOp\Mod
+				|| $node instanceof Node\Expr\AssignOp\Mod
+			) {
+				if ($stringType instanceof ConstantScalarType && $stringType->getValue() === '1') {
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php:1061]: This logic is only going to be executed for ConstantStringType. But the same logic works for `1.0` as well: https://3v4l.org/CYeUe

I'd rather be if you called `->toInteger()` and checked ConstantIntegerType afterwards.

File: src/Analyser/MutatingScope.php [hunk 2]
```diff
@@ -1051,12 +1051,24 @@ private function resolveType(Expr $node): Type
...
+			$rightScalarTypes = TypeUtils::getConstantScalars($rightType->toNumber());
+			foreach ($rightScalarTypes as $scalarType) {
 
-			$rightTypes = TypeUtils::getConstantScalars($this->getType($right)->toNumber());
-			foreach ($rightTypes as $rightType) {
 				if (
-					$rightType->getValue() === 0
-					|| $rightType->getValue() === 0.0
+					$scalarType->getValue() === 0
+					|| $scalarType->getValue() === 0.0
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php:1071]: What about `'0'` here? Again, you can try `->toInteger()`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Analyser/MutatingScope.php (hunk 2 above)
Comment (staabm) [reply src/Analyser/MutatingScope.php:1071]: this code was existant before this PR and covers a different use-case.
should I change it with this PR nevertheless , or better do so in a different PR?

Comment (ondrejmirtes) [reply src/Analyser/MutatingScope.php:1071]: You can do it here :)

Comment (staabm) [reply src/Analyser/MutatingScope.php:1071]: newly added testcases in 578e96d show that division by '0' and other falsy values work as expected.
so IMO we just leave this as is, right?

**Comment by ondrejmirtes** [comment]
Thank you.

**Pull request merged**